src/ros_tcp_endpoint/tcp_sender.py

```diff
@@ -19,108 +19,103 @@
-            return
-
-        serialized_message = ClientThread.serialize_message(topic, request)
+        request_bytes = BytesIO()
+        request.serialize(request_bytes)
+        result = [None]
+        condition = threading.Condition()
         
-        try:
-            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            s.settimeout(self.timeout_on_connect)
-            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-            s.connect((self.unity_ip, self.unity_port))
-            s.settimeout(self.timeout_on_send)
-            s.sendall(serialized_message)
-
-            destination, data = ClientThread.read_message(s)
+        with self.srv_lock:
+            srv_id = self.next_srv_id
+            self.next_srv_id+=1
+            self.services_waiting[srv_id] = (condition, result)
+        
+        payload = request_bytes.getvalue()
+        srv_message = RosUnitySrvMessage(srv_id, True, topic, payload)
+        serialized_message = ClientThread.serialize_message('__srv', srv_message)
+        self.queue.put(serialized_message)
 
-            response = service_class._response_class().deserialize(data)
+        # rospy starts a new thread for each service request,
+        # so it won't break anything if we suspend it while waiting for the response
+        with condition:
+            condition.wait()
+        
+        response = service_class._response_class().deserialize(result[0])
+        return response
+    
+    def send_unity_service_response(self, srv_id, data):
+        with self.srv_lock:
+            (condition, result) = self.services_waiting[srv_id]
+            result[0] = data
+            del self.services_waiting[srv_id]
+        
+        with condition:
+            condition.notify()
 
-            s.close()
-            return response
-        except Exception as e:
-            rospy.loginfo("Exception {}".format(e))
+    def start_sender(self, conn, reader_halt, sender_halt):
+        sender_thread = threading.Thread(target=self.sender_loop, args=(conn, self.sender_id, reader_halt, sender_halt))
+        self.sender_id += 1
+        # Exit the server thread when the main thread terminates
+        sender_thread.daemon = True
+        sender_thread.start()
  
-    def sender_loop(self):
+    def sender_loop(self, conn, tid, reader_halt, sender_halt):
         s = None
-        idletimeout = 0
         
         while True:
-            item = self.queue.get()
+            try:
+                item = self.queue.get(timeout=0.5)
+            except Empty:
+                # ideally we'd just wait on the queue, but we also need to check frequently for the connection being closed 
+                # (otherwise we might take a message and fail to send it)
+                if sender_halt[0]:
+                    return
+                continue
+
+            if sender_halt[0]:
+                return
```

Comment:
If you change the while loop, you shouldn't need either of these `if sender_halt[0]` right?  btw the `sender_halt` bool might be another candidate for encapsulation in a descriptively named class (`ControlSignals`?) instead of a List